expo/expo
# fix(cli): add `EMFILE` error handling for MacOS to clarify why Metro crashed

# Why

Fixes ENG-13486

This adds basic error handling for `EMFILE` errors on MacOS, and outputs two separate messages.

1. When Watchman is not installed
![watchman-not-installed](https://github.com/user-attachments/assets/0ec2d89a-3183-4cc1-b96b-cf555fd3ae46)

2. When Watchman is in a bad state
![watchman-bad-state](https://github.com/user-attachments/assets/bcbed353-d5e3-43af-a122-56481ed0932d)


## What do you mean "watchman is in a bad state"?

Well, it turns out that the [command that Metro is using to determine if watchman is installed](https://github.com/facebook/metro/blob/f2a130c324c50b5c586b2d7e36bf6be7a8e49795/packages/metro-file-map/src/lib/checkWatchmanCapabilities.js#L24-L29) and can be used, might cause a [segfault](https://github.com/facebook/watchman/issues/400).

<img width="760" alt="image" src="https://github.com/user-attachments/assets/c1188484-1881-45b1-8eee-e594699d9178">

> You can test this by running: `watchman shutdown-server && watchman list-capabilities --output-encoding=json --no-pretty --no-spawn`

It's basically related to the `--no-spawn`, when running the watchman server, everything works fine.

<img width="473" alt="image" src="https://github.com/user-attachments/assets/b6284418-f696-428f-81de-6cd98bc60ccb">

<img width="552" alt="image" src="https://github.com/user-attachments/assets/c5bd89d1-162d-4db2-a0d3-89dd58f359a1">

So, whenever we detect the EMFILE on MacOS, and we detect Watchman is installed -- we try to restart this server and [reset all file watchers](https://github.com/facebook/metro/blob/f2a130c324c50b5c586b2d7e36bf6be7a8e49795/docs/Troubleshooting.md?plain=1#L8). On the next start, this should get watchman in a working state and Metro to use watchman instead.

# How

- Added error handling through `process.on('uncaughtException', ..)`
- Added watchman restart / detection commands
- Added log to install watchman
- Added log to troubleshoot watchman, following Metro's own [watchman troubleshooting guide](https://github.com/facebook/metro/blob/f2a130c324c50b5c586b2d7e36bf6be7a8e49795/docs/Troubleshooting.md?plain=1#L8)

# Test Plan

On my mac, this reproduces it reliably:

- `brew install watchman`
- `npx create-expo@latest ./test-expo-emfile --template default@next`
- `cd ./test-expo-emfile`
- `watchman shutdown-server` (to trigger the bad state)
- `npx expo start`
  - Should see the bad state warning

To test the install watchman warning:

- `brew remove watchman`
- `npx expo start`
  - Should see the install watchman warning

# Checklist
